RHTAPINST-39: Embedded Helm Charts

otaviof opened the pull request:
Embedding the Helm Charts and other resources needed for the installation process in the final executable `rhtap-cli`. The flag `--embedded` is enabled by default, and can be set to `false` in order to use local installer resources.

Golang `embed` doesn't allow including symbolic links and files starting with `_` (underscore), thus the approach chosen for this PR is using a tarball from the `installer` directory and navigating through it as a regular `fs.FS` interface.

```bash
bin/rhtap-cli deploy --embedded=true
```

It's also possible to list and extract embedded contents using:

```bash
bin/rhtap-cli installer --list

mkdir /var/tmp/rhtap-cli-resources
bin/rhtap-cli installer --extract /var/tmp/rhtap-cli-resources
```

otaviof committed:
chore: Installer Directory

Moving Helm Charts, configuration and scripts into a dedicated folder
named `installer`, to store all resources needed for the `rhtap-cli`
deployment process.

otaviof committed:
chore: Vendoring Dependencies

otaviof committed:
feat: Embedded Resources

Creating insaller executable with all the required resources for the
deployment process.

Roming22 commented on CONTRIBUTING.md:
````diff
@@ -16,6 +16,17 @@ After you have cloned the repository and installed the requirements, you can sta
 make
 ```
 
+The resources needed for the installer are stored in the `installer` directory, you can find the `config.yaml` file there. Before the application is built, the contents of the installer directory are package into a tarball and embedded into the application binary.
````
*are packaged

Roming22 commented on Dockerfile:
```diff
@@ -8,13 +8,13 @@ USER root
 
 WORKDIR /workdir/rhtap-cli
 
-COPY charts/ ./charts/
```
Why not also keep the resources in the image so users can easily customize the deployment if necessary?

otaviof replied:
These resources are on the `installer` directory, copied into the image accordingly.

Roming22 commented on README.md:
````diff
@@ -102,32 +102,34 @@ Please refer to the [CONTRIBUTING.md](CONTRIBUTING.md) file for more information
 ```
   
-5.Finally run the below command to proceed with RHTAP deployment. 
+5. Finally run the below command to proceed with RHTAP deployment. 
````
Should the example use `--embedded=false --config=config.yaml`, as the user might have modified the config.yaml?

otaviof replied:
The `config.yaml` isn't embedded, to this file is still required. Only charts and scripts are embedded currently.

Roming22 commented on pkg/subcmd/deploy.go:
```diff
@@ -36,6 +35,9 @@ applied, on the attribute 'rhtapCLI.dependencies[]'.
 The platform configuration is rendered from the values template file
 (--values-template), this configuration payload is given to all Helm charts.
+
+The installer resources are embedded on the executable, these resources are
```
*embedded in

Roming22 commented on pkg/subcmd/template.go:
```diff
@@ -44,6 +42,9 @@ chart directory, optional.
 Additionally, the '--debug' flag should be used to display the raw values template
 payload, regardless of whether it is valid YAML or not.
+
+The installer resources are embedded on the executable, these resources are
```
*embedded in

Roming22 commented:
@otaviof You need to update `yamllint.yaml`, and take care of sonarcloud and snyk before it can be merged.

Roming22 commented:
@otaviof PR is still broken.

otaviof commented:
> @otaviof You need to update `yamllint.yaml`, and take care of sonarcloud and snyk before it can be merged.

The SonarCloud alerts have been triggered after moving files into the `installer` directory, have we ignored these alerts before? Or, perhaps, this is a new static code analysis? Well, anyway, given this PR is not modifying these files, we should have a dedicated PR to fix the issues. We can ignore them for the time being, but fix right after.

otaviof committed:
chore: SonarCloud Fixes

Fixing style issues reported by SonarCloud.

otaviof commented:
> > @otaviof You need to update `yamllint.yaml`, and take care of sonarcloud and snyk before it can be merged.
> 
> The SonarCloud alerts have been triggered after moving files into the `installer` directory, have we ignored these alerts before? Or, perhaps, this is a new static code analysis? Well, anyway, given this PR is not modifying these files, we should have a dedicated PR to fix the issues. We can ignore them for the time being, but fix right after.

@Roming22, I've included on more commit to fix the styling part of SonarCloud alerts, and documented the rest using the web interface. Please consider.

Roming22 commented:
/retest

Roming22 commented:
/lgtm

otaviof committed:
chore: Container Image

Placing the executable on the `${PATH}` scope, and using the latest UBI
9.4 image.

otaviof commented:
/retest

otaviof commented:
/retest

otaviof commented:
/retest

otaviof commented:
/retest

otaviof commented:
/retest

flacatus commented:
due that we cannot test yet in konflux task changes we test manually changing the task to otavio fork and the test pass!

![Screenshot from 2024-09-11 10-27-12](https://github.com/user-attachments/assets/110051f7-7e3c-43d9-8a4d-c1a82da6429b)

otaviof commented:
> due that we cannot test yet in konflux task changes we test manually changing the task to otavio fork and the test pass!
> 
> ![Screenshot from 2024-09-11 10-27-12](https://private-user-images.githubusercontent.com/59865209/366364736-110051f7-7e3c-43d9-8a4d-c1a82da6429b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNDM1NDEsIm5iZiI6MTcyNjA0MzI0MSwicGF0aCI6Ii81OTg2NTIwOS8zNjYzNjQ3MzYtMTEwMDUxZjctN2UzYy00M2Q5LThhNGQtYzFhODJkYTY0MjliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTExVDA4MjcyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0ZTEwOGEwOGJhZjc2OGNhNDExYjM5MTgxZThlZjFkMTAxOWJkZWNiOTg1ZWI0ZjEyMzMzZWQ5ZGQxYzJkY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.G9Ltk2b0pwOJwZKq-JwKNOYVttuxQsx_WRp_VZ5-Fu0)

I'm going to remove the last commit, with the temporary settings, and then we can merge this afterwards. Thanks!

Roming22 commented:
/cherry-pick release-1.2
